### PR TITLE
raft: implemented leader lease when quorum check is on

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
port leader lease from coreos/etcd
https://github.com/coreos/etcd/pull/5468
https://github.com/coreos/etcd/pull/5809

### CODE REVIEW

File: src/raft/raft.rs [hunk 1]
```diff
@@ -611,7 +614,7 @@ impl<T: Storage> Raft<T> {
         info!("{} became leader at term {}", self.tag, self.term);
     }
 
-    fn campaign(&mut self) {
+    fn campaign(&mut self, campaign_type: Vec<u8>) {
```
Comment (siddontang) [inline src/raft/raft.rs]: why not using &[u8]?

File: src/raft/raft.rs [hunk 2]
```diff
@@ -677,6 +683,29 @@ impl<T: Storage> Raft<T> {
...
             let leader_id = if m.get_msg_type() == MessageType::MsgRequestVote {
+
+                let force = m.get_context() == CAMPAIGN_TRANSFER;
+                let in_lease = self.check_quorum && self.state != StateRole::Candidate &&
+                               self.election_elapsed < self.election_timeout;
+                if !force && in_lease {
+                    // if a server receives ReqeustVote request within the minimum election
+                    // timeout of hearing from a current leader, it does not update its term
+                    // or grant its vote
+                    info!("{} [logterm: {}, index: {}, vote: {}] ignored vote from {} [logterm: \
```
Comment (ngaut) [inline src/raft/raft.rs]: Also print  self.tag.

File: src/raft/raft.rs [hunk 3]
```diff
@@ -677,6 +683,29 @@ impl<T: Storage> Raft<T> {
             // local message
         } else if m.get_term() > self.term {
             let leader_id = if m.get_msg_type() == MessageType::MsgRequestVote {
+
```
Comment (ngaut) [inline src/raft/raft.rs]: Remove blank line.

File: tests/test_raft.rs [hunk 4]
```diff
@@ -1358,6 +1358,164 @@ fn test_leader_stepdown_when_quorum_lost() {
...
+
+    // prevent campaigning from b
+    nt.peers.get_mut(&2).unwrap().set_randomized_election_timeout(b_election_timeout + 1);
+    for _ in 0..b_election_timeout {
+        nt.peers.get_mut(&2).unwrap().tick();
+    }
+    nt.send(vec![new_message(1, 1, MessageType::MsgHup, 0)]);
+
+    assert_eq!(nt.peers.get(&1).unwrap().state, StateRole::Leader);
+
```
Comment (ngaut) [inline tests/test_raft.rs]: Remove blank line

File: tests/test_raft.rs [hunk 5]
```diff
@@ -1358,6 +1358,164 @@ fn test_leader_stepdown_when_quorum_lost() {
...
+    let b_election_timeout = nt.peers.get(&2).unwrap().get_election_timeout();
+
+    // Letting b's electionElapsed reach to timeout so that it can vote for a
+    for _ in 0..b_election_timeout {
+        nt.peers.get_mut(&2).unwrap().tick();
+    }
+    nt.send(vec![new_message(1, 1, MessageType::MsgHup, 0)]);
+
+    assert_eq!(nt.peers.get(&1).unwrap().state, StateRole::Leader);
+
```
Comment (ngaut) [inline tests/test_raft.rs]: Remove blank line.

File: tests/test_raft.rs [hunk 6]
```diff
@@ -1358,6 +1358,164 @@ fn test_leader_stepdown_when_quorum_lost() {
...
+    for _ in 0..a_election_timeout {
+        nt.peers.get_mut(&1).unwrap().tick();
+    }
+    for _ in 0..b_election_timeout {
+        nt.peers.get_mut(&2).unwrap().tick();
+    }
+    nt.send(vec![new_message(3, 3, MessageType::MsgHup, 0)]);
+
+    assert_eq!(nt.peers.get(&1).unwrap().state, StateRole::Follower);
+
```
Comment (ngaut) [inline tests/test_raft.rs]: Remove blank line

File: tests/test_raft.rs [hunk 7]
```diff
@@ -1358,6 +1358,164 @@ fn test_leader_stepdown_when_quorum_lost() {
...
+    for _ in 0..b_election_timeout {
+        nt.peers.get_mut(&2).unwrap().tick();
+    }
+    nt.send(vec![new_message(1, 1, MessageType::MsgHup, 0)]);
+
+    nt.isolate(1);
+    nt.send(vec![new_message(3, 3, MessageType::MsgHup, 0)]);
+
+    assert_eq!(nt.peers.get(&2).unwrap().state, StateRole::Follower);
+
```
Comment (ngaut) [inline tests/test_raft.rs]: Remove blank line.

File: tests/test_raft.rs [hunk 8]
```diff
@@ -1358,6 +1358,164 @@ fn test_leader_stepdown_when_quorum_lost() {
...
+    assert_eq!(nt.peers.get(&3).unwrap().state, StateRole::Candidate);
+
```
Comment (ngaut) [inline tests/test_raft.rs]: Remove blank line.

File: tests/test_raft.rs [hunk 9]
```diff
@@ -1358,6 +1358,164 @@ fn test_leader_stepdown_when_quorum_lost() {
...
+    assert_eq!(nt.peers.get(&3).unwrap().term,
+               nt.peers.get(&2).unwrap().term + 1);
+
+    // Vote again for safety
+    nt.send(vec![new_message(3, 3, MessageType::MsgHup, 0)]);
+
+    assert_eq!(nt.peers.get(&2).unwrap().state, StateRole::Follower);
+
```
Comment (ngaut) [inline tests/test_raft.rs]: Remove blank line.

File: tests/test_raft.rs [hunk 10]
```diff
@@ -1358,6 +1358,164 @@ fn test_leader_stepdown_when_quorum_lost() {
...
+    assert_eq!(nt.peers.get(&3).unwrap().state, StateRole::Candidate);
+
```
Comment (ngaut) [inline tests/test_raft.rs]: Remove blank line.

File: tests/test_raft.rs [hunk 11]
```diff
@@ -1358,6 +1358,164 @@ fn test_leader_stepdown_when_quorum_lost() {
...
+               nt.peers.get(&2).unwrap().term + 2);
+
+    nt.recover();
+    let mut msg = new_message(1, 3, MessageType::MsgHeartbeat, 0);
+    msg.set_term(nt.peers.get(&1).unwrap().term);
+    nt.send(vec![msg]);
+
+    // Disrupt the leader so that the stuck peer is freed
+    assert_eq!(nt.peers.get(&1).unwrap().state, StateRole::Follower);
+
```
Comment (ngaut) [inline tests/test_raft.rs]: Remove blank line.

File: tests/test_raft.rs [hunk 12]
```diff
@@ -1358,6 +1358,164 @@ fn test_leader_stepdown_when_quorum_lost() {
...
+    assert_eq!(nt.peers.get(&2).unwrap().promotable(), false);
+
+    let b_election_timeout = nt.peers.get(&2).unwrap().get_election_timeout();
+    for _ in 0..b_election_timeout {
+        nt.peers.get_mut(&2).unwrap().tick();
+    }
+    nt.send(vec![new_message(1, 1, MessageType::MsgHup, 0)]);
+
+    assert_eq!(nt.peers.get(&1).unwrap().state, StateRole::Leader);
+
```
Comment (ngaut) [inline tests/test_raft.rs]: Remove blank line.

File: tests/test_raft.rs [hunk 13]
```diff
@@ -1358,6 +1358,164 @@ fn test_leader_stepdown_when_quorum_lost() {
...
+    assert_eq!(nt.peers.get(&2).unwrap().state, StateRole::Follower);
+
```
Comment (ngaut) [inline tests/test_raft.rs]: Remove blank line.

File: tests/test_raft.rs [hunk 14]
```diff
@@ -1969,6 +2127,38 @@ fn test_leader_transfer_to_uptodate_node() {
...
+    assert_eq!(nt.peers.get(&1).unwrap().leader_id, 1);
+
+    // Transfer leadership to 2.
+    nt.send(vec![new_message(2, 1, MessageType::MsgTransferLeader, 0)]);
+    check_leader_transfer_state(nt.peers.get(&1).unwrap(), StateRole::Follower, 2);
+
+    // After some log replication, transfer leadership back to 1.
+    nt.send(vec![new_message(1, 1, MessageType::MsgPropose, 1)]);
+
```
Comment (ngaut) [inline tests/test_raft.rs]: Remove blank line.

File: tests/test_raft.rs [hunk 15]
```diff
@@ -1969,6 +2127,38 @@ fn test_leader_transfer_to_uptodate_node() {
...
+    nt.send(vec![new_message(1, 2, MessageType::MsgTransferLeader, 0)]);
+
```
Comment (ngaut) [inline tests/test_raft.rs]: Remove blank line.

File: src/raft/raft.rs [hunk 16]
```diff
@@ -306,6 +308,11 @@ impl<T: Storage> Raft<T> {
         self.prs.len() / 2 + 1
     }
 
+    // for leader lease test
```
Comment (ngaut) [inline src/raft/raft.rs]: for testing leader lease

**Comment by zhangjinpeng1987** [comment]
@siddontang @BusyJay @disksing @ngaut PTAL

File: src/raft/raft.rs [hunk 17]
```diff
@@ -677,6 +683,30 @@ impl<T: Storage> Raft<T> {
...
+                let in_lease = self.check_quorum && self.state != StateRole::Candidate &&
+                               self.election_elapsed < self.election_timeout;
+                if !force && in_lease {
+                    // if a server receives ReqeustVote request within the minimum election
+                    // timeout of hearing from a current leader, it does not update its term
+                    // or grant its vote
+                    info!("{} [logterm: {}, index: {}, vote: {}, tag: {}] ignored vote from {} \
+                           [logterm: {}, index: {}] at term {}: lease is not expired (remaining \
+                           ticks: {})",
+                          self.id,
```
Comment (siddontang) [inline src/raft/raft.rs]: use self.tag instead

Comment (siddontang) [inline src/raft/raft.rs]: we use tag at first.

**Comment by siddontang** [comment]
etcd has same problem https://github.com/coreos/etcd/issues/6068

**Comment by zhangjinpeng1987** [comment]
@siddontang @BusyJay @ngaut @disksing PTAL

**Comment by siddontang** [comment]
LGTM
But if https://github.com/coreos/etcd/pull/6070 updates, we should update it later.

File: src/raft/raft.rs [hunk 18]
```diff
@@ -633,6 +636,9 @@ impl<T: Storage> Raft<T> {
             let mut m = new_message(id, MessageType::MsgRequestVote, None);
             m.set_index(self.raft_log.last_index());
             m.set_log_term(self.raft_log.last_term());
+            if campaign_type == CAMPAIGN_TRANSFER {
+                m.set_context(Vec::from(campaign_type));
```
Comment (BusyJay) [inline src/raft/raft.rs]: `compaign_type.to_vec()`.

File: src/raft/raft.rs [hunk 19]
```diff
@@ -677,6 +683,29 @@ impl<T: Storage> Raft<T> {
             // local message
         } else if m.get_term() > self.term {
             let leader_id = if m.get_msg_type() == MessageType::MsgRequestVote {
+                let force = m.get_context() == CAMPAIGN_TRANSFER;
+                let in_lease = self.check_quorum && self.state != StateRole::Candidate &&
```
Comment (BusyJay) [inline src/raft/raft.rs]: I think we should check whether `self.leader_id` is `INVALID_ID`.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**